src/vm/analysis/read_only_checker/mod.rs

```diff
@@ -77,97 +93,132 @@ impl<'a, 'b> ReadOnlyChecker<'a, 'b> {
-    ///   (2) if valid, returns whether or not they are read only.
-    /// Note that because of (1), this function _cannot_ short-circuit on read-only.
-    fn check_read_only(&mut self, expr: &SymbolicExpression) -> CheckResult<bool> {
-        match expr.expr {
+    /// Checks that a function with signature `signature` and body `body` is read-only.
+    ///
+    /// This is used to check the function definitions `PrivateFunction`, `PublicFunction`
+    /// and `ReadOnlyFunction`.
+    ///
+    /// Returns a pair of 1) the function name defined, and 2) a `bool` indicating whether the function
+    /// is read-only.
+    fn check_define_function(
+        &mut self,
+        signature: &[SymbolicExpression],
+        body: &SymbolicExpression,
+    ) -> CheckResult<(ClarityName, bool)> {
+        let function_name = signature
+            .get(0)
+            .ok_or(CheckErrors::DefineFunctionBadSignature)?
+            .match_atom()
+            .ok_or(CheckErrors::BadFunctionName)?;
+
+        let is_read_only = self.check_atomic_expression_is_read_only(body)?;
+
+        Ok((function_name.clone(), is_read_only))
+    }
+
+    /// Checks an atomic `expression` to determine whether it is read-only correct.
+    /// An atomic expression is one that does not need to be parsed into multiple expressions.
+    ///
+    /// Returns `true` iff the expression is read-only.
+    fn check_atomic_expression_is_read_only(
+        &mut self,
+        expression: &SymbolicExpression,
+    ) -> CheckResult<bool> {
+        match expression.expr {
             AtomValue(_) | LiteralValue(_) | Atom(_) | TraitReference(_, _) | Field(_) => Ok(true),
-            List(ref expression) => self.check_function_application_read_only(expression),
+            List(ref expression) => self.check_expression_application_is_read_only(expression),
         }
     }
 
-    /// Checks all of the supplied symbolic expressions
-    ///   (1) for whether or not they are valid with respect to read-only requirements.
-    ///   (2) if valid, returns whether or not they are read only.
-    /// Note that because of (1), this function _cannot_ short-circuit on read-only.
-    fn check_all_read_only(&mut self, expressions: &[SymbolicExpression]) -> CheckResult<bool> {
+    /// Checks each expression in `expressions` to determine whether each uses only
+    /// read-only operations.
+    ///
+    /// Returns `true` iff all expressions are read-only.
+    fn check_each_expression_is_read_only(
+        &mut self,
+        expressions: &[SymbolicExpression],
+    ) -> CheckResult<bool> {
         let mut result = true;
-        for expr in expressions.iter() {
-            let expr_read_only = self.check_read_only(expr)?;
+        for expression in expressions.iter() {
+            let expr_read_only = self.check_atomic_expression_is_read_only(expression)?;
             result = result && expr_read_only;
```

Comment:
Do you understand why we don't do an early return is `expr_read_only` is ever false?

Reply:
Not sure.. @kantai ? Should we return early?

Reply:
Yes -- this method shouldn't return early, because it's responsible not only for checking that a given expression isn't read-only, but also enforcing a read-only constraint (see the thrown CheckErrors)

Reply:
ok.. i added a note too